# 01: Audio Encoder Training (A1, A2, A3)

Train three MuQ domain adaptation strategies on Thunder Compute:
- **A1**: MuQ + LoRA multi-task (MuQLoRAModel)
- **A2**: Staged domain adaptation (MuQStagedModel) - Stage 1: self-supervised, Stage 2: supervised
- **A3**: Full unfreeze with gradual layer unfreezing (MuQFullUnfreezeModel)

All models use pre-extracted MuQ embeddings and PercePiano labels with 4-fold cross-validation.

---

## 1. Setup

In [1]:
import subprocess
import sys
import os
import torch
if torch.cuda.is_available():
    os.environ['CUBLAS_WORKSPACE_CONFIG'] = ':4096:8'

In [2]:
import sys
from pathlib import Path

IS_CLOUD = Path('/workspace').exists()

if IS_CLOUD:
    get_ipython().system('git clone https://github.com/Jai-Dhiman/crescendAI.git /workspace/crescendai')
    get_ipython().run_line_magic('cd', '/workspace/crescendai/model')
    get_ipython().system('curl -LsSf https://astral.sh/uv/install.sh | sh')
    get_ipython().system(f'uv pip install -e . --python {sys.executable}')
    get_ipython().system('rclone sync gdrive:crescendai_data/model_improvement/data/percepiano_cache ./data/percepiano_cache --progress')
    get_ipython().system('rclone sync gdrive:crescendai_data/model_improvement/data/competition_cache ./data/competition_cache --progress')
    get_ipython().system('rclone sync gdrive:crescendai_data/model_improvement/data/maestro_cache ./data/maestro_cache --progress')
    get_ipython().system("rclone copy gdrive:crescendai_data/model_improvement/data/composite_labels ../data/composite_labels --progress 2>/dev/null || echo 'composite_labels: using local copy'")
else:
    # cd to model/ so all relative paths (data/, src/) work consistently
    _nb_dir = Path.cwd()
    _model_dir = _nb_dir
    while _model_dir.name != 'model' and _model_dir != _model_dir.parent:
        _model_dir = _model_dir.parent
    if _model_dir.name == 'model':
        get_ipython().run_line_magic('cd', str(_model_dir))

DATA_DIR = Path('data')
CHECKPOINT_DIR = Path('data/checkpoints/model_improvement')

CHECKPOINT_DIR.mkdir(parents=True, exist_ok=True)
print(f'DATA_DIR: {DATA_DIR.resolve()}')
print(f'CHECKPOINT_DIR: {CHECKPOINT_DIR.resolve()}')
print(f'Running on: {"cloud" if IS_CLOUD else "local"}')

/Users/jdhiman/Documents/crescendai/model
DATA_DIR: /Users/jdhiman/Documents/crescendai/model/data
CHECKPOINT_DIR: /Users/jdhiman/Documents/crescendai/model/data/checkpoints/model_improvement
Running on: local


In [3]:
import json
import numpy as np
import torch
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping
from torch.utils.data import DataLoader
from functools import partial

sys.path.insert(0, 'src')

from model_improvement.audio_encoders import MuQLoRAModel, MuQStagedModel, MuQFullUnfreezeModel
from model_improvement.training import train_model, find_checkpoint, detect_accelerator_config, upload_checkpoint
from model_improvement.data import (
    PairedPerformanceDataset, AugmentedEmbeddingDataset, AudioSegmentDataset,
    MaestroContrastiveDataset, stage1_collate_fn,
    audio_pair_collate_fn, multi_task_collate_fn,
)
from model_improvement.augmentation import AudioAugmentor
from model_improvement.taxonomy import load_composite_labels, NUM_DIMS, DIMENSIONS
from model_improvement.evaluation import (
    evaluate_model, aggregate_folds, run_robustness_check,
    select_winner, ROBUSTNESS_VETO_THRESHOLD,
)
from model_improvement.metrics import format_comparison_table
import matplotlib.pyplot as plt

## 2. Load Data

In [4]:
cache_dir = DATA_DIR / 'percepiano_cache'

# Load 6-dim composite labels (teacher-grounded taxonomy)
composite_path = DATA_DIR / 'composite_labels' / 'composite_labels.json'
labels_raw = load_composite_labels(composite_path)
labels = {k: v.tolist() for k, v in labels_raw.items()}
print(f'Loaded {len(labels)} composite labels ({NUM_DIMS} dims: {DIMENSIONS})')

emb_path = cache_dir / 'muq_embeddings.pt'
if emb_path.exists():
    embeddings = torch.load(emb_path, map_location='cpu', weights_only=True)
    print(f'Loaded {len(embeddings)} pre-computed embeddings')
else:
    print('muq_embeddings.pt not found -- extracting from audio (one-time, ~20 min on A100)...')
    import gc
    from audio_experiments.extractors.muq import MuQExtractor

    audio_dir = DATA_DIR / 'percepiano_audio'
    if not audio_dir.exists() or not any(audio_dir.glob('*.wav')):
        raise FileNotFoundError(
            f'No wav files at {audio_dir}. Sync percepiano_audio/ from GDrive first.'
        )

    extractor = MuQExtractor(cache_dir=cache_dir / '_muq_file_cache')
    embeddings = {}
    missing = []
    keys = sorted(labels.keys())
    for i, key in enumerate(keys, 1):
        wav = audio_dir / f'{key}.wav'
        if not wav.exists():
            missing.append(key)
            continue
        embeddings[key] = extractor.extract_from_file(wav)
        if i % 100 == 0:
            print(f'  [{i}/{len(keys)}] extracted')

    del extractor
    gc.collect()
    torch.cuda.empty_cache()

    if missing:
        print(f'WARNING: {len(missing)} wav files not found')

    torch.save(embeddings, emb_path)
    print(f'Saved {len(embeddings)} embeddings to {emb_path}')

    # Cache to GDrive so future runs skip extraction
    subprocess.run([
        'rclone', 'copy', str(emb_path),
        'gdrive:crescendai_data/model_improvement/data/percepiano_cache/',
        '--progress',
    ], check=True)

print(f'Labels: {len(labels)}, Embeddings: {len(embeddings)}')
sample_key = next(iter(embeddings))
print(f'Sample embedding shape: {embeddings[sample_key].shape}')

Loaded 1202 composite labels (6 dims: ['dynamics', 'timing', 'pedaling', 'articulation', 'phrasing', 'interpretation'])
Loaded 1201 pre-computed embeddings
Labels: 1202, Embeddings: 1201
Sample embedding shape: torch.Size([426, 1024])


In [5]:
with open(cache_dir / 'folds.json') as f:
    folds = json.load(f)

print(f'Number of folds: {len(folds)}')
for i, fold in enumerate(folds):
    print(f'  Fold {i}: {len(fold["train"])} train, {len(fold["val"])} val')

Number of folds: 4
  Fold 0: 758 train, 263 val
  Fold 1: 787 train, 234 val
  Fold 2: 766 train, 255 val
  Fold 3: 752 train, 269 val


In [6]:
# Build piece mapping from labels metadata
# Each label key encodes piece info; group recordings by piece
with open(cache_dir / 'piece_mapping.json') as f:
    piece_to_keys = json.load(f)

print(f'Pieces with multiple performances: {len(piece_to_keys)}')
total_keys = sum(len(v) for v in piece_to_keys.values())
print(f'Total recordings in piece mapping: {total_keys}')

Pieces with multiple performances: 61
Total recordings in piece mapping: 964


## 3. Training Utilities

In [7]:
# On macOS (spawn), collate_fn closures pickle the full embeddings dict
# (~2 GB) into each worker -- use 0 workers locally to avoid OOM.
_NUM_WORKERS = 4 if torch.cuda.is_available() else 0
_PIN_MEMORY = torch.cuda.is_available()

LOW_MEMORY = not torch.cuda.is_available()
_BATCH_SIZE = 4 if LOW_MEMORY else 16
_ACCUM_BATCHES = 4 if LOW_MEMORY else 1

print(f'[memory config] LOW_MEMORY={LOW_MEMORY}  batch_size={_BATCH_SIZE}  accum={_ACCUM_BATCHES}  workers={_NUM_WORKERS}')


def build_dataloaders(fold, labels, piece_to_keys, embeddings, cache_dir, batch_size=_BATCH_SIZE, include_augmented=False):
    """Build train/val DataLoaders for a given fold."""
    collate_fn = partial(audio_pair_collate_fn, embeddings=embeddings)

    if include_augmented:
        # Self-supervised: per-segment dataset with clean + augmented embeddings
        train_ds = AudioSegmentDataset(
            embeddings=embeddings,
            piece_to_keys=piece_to_keys,
            keys=fold['train'],
        )
        val_ds = AudioSegmentDataset(
            embeddings=embeddings,
            piece_to_keys=piece_to_keys,
            keys=fold['val'],
        )
        collate_fn = multi_task_collate_fn
    else:
        # Supervised: paired dataset with audio collate
        train_ds = PairedPerformanceDataset(
            cache_dir=cache_dir,
            labels=labels,
            piece_to_keys=piece_to_keys,
            keys=fold['train'],
        )
        val_ds = PairedPerformanceDataset(
            cache_dir=cache_dir,
            labels=labels,
            piece_to_keys=piece_to_keys,
            keys=fold['val'],
        )

    train_loader = DataLoader(
        train_ds, batch_size=batch_size, shuffle=True,
        collate_fn=collate_fn, num_workers=_NUM_WORKERS, pin_memory=_PIN_MEMORY,
    )
    val_loader = DataLoader(
        val_ds, batch_size=batch_size, shuffle=False,
        collate_fn=collate_fn, num_workers=_NUM_WORKERS, pin_memory=_PIN_MEMORY,
    )
    return train_loader, val_loader

[memory config] LOW_MEMORY=True  batch_size=4  accum=4  workers=0


## 4. Train A1: MuQ + LoRA

In [8]:
A1_CONFIG = {
    'input_dim': 1024,
    'hidden_dim': 512,
    'num_labels': NUM_DIMS,
    'learning_rate': 3e-5,
    'weight_decay': 1e-5,
    'temperature': 0.07,
    'lambda_contrastive': 0.3,
    'lambda_regression': 0.5,
    'lambda_invariance': 0.1,
    'warmup_epochs': 5,
    'max_epochs': 200,
    'use_pretrained_muq': False,  # Using pre-extracted embeddings
}

print('Training A1: MuQ + LoRA Multi-Task')
print('=' * 50)

a1_trainers = []
a1_models = []
for fold_idx, fold in enumerate(folds):
    print(f'\nFold {fold_idx}/{len(folds)-1}')

    existing = find_checkpoint(CHECKPOINT_DIR, 'A1', fold_idx)
    if existing:
        print(f'  Loading existing checkpoint: {existing.name}')
        model = MuQLoRAModel.load_from_checkpoint(existing, **A1_CONFIG)
        a1_models.append(model)
        a1_trainers.append(None)
        continue

    model = MuQLoRAModel(**A1_CONFIG)
    train_loader, val_loader = build_dataloaders(
        fold, labels, piece_to_keys, embeddings, cache_dir,
        batch_size=_BATCH_SIZE,
    )

    trainer = train_model(
        model, train_loader, val_loader, 'A1', fold_idx,
        checkpoint_dir=CHECKPOINT_DIR,
        max_epochs=A1_CONFIG['max_epochs'],
        patience=10,
        accumulate_grad_batches=_ACCUM_BATCHES,
    )
    a1_trainers.append(trainer)
    a1_models.append(trainer.lightning_module)

    best_val = trainer.callback_metrics.get('val_loss', float('inf'))
    best_acc = trainer.callback_metrics.get('val_pairwise_acc', 0.0)
    print(f'  Best val_loss={best_val:.4f}, val_pairwise_acc={best_acc:.4f}')

Training A1: MuQ + LoRA Multi-Task

Fold 0/3
  Loading existing checkpoint: epoch=8-val_loss=0.4879.ckpt

Fold 1/3
  Loading existing checkpoint: epoch=5-val_loss=0.5473.ckpt

Fold 2/3
  Loading existing checkpoint: epoch=6-val_loss=0.4325.ckpt

Fold 3/3
  Loading existing checkpoint: epoch=5-val_loss=0.5450.ckpt


## 5. Train A2: Staged Domain Adaptation

### Note: A2 pretraining stays on PercePiano only (for now)

Unlike the symbolic encoders (02_symbolic_training.ipynb) which now pretrain on a
~14K MIDI corpus (MAESTRO + ATEPP + ASAP + PercePiano), the audio encoders here
continue to pretrain only on PercePiano data. The rationale:

1. **MuQ already has strong audio priors.** MuQ is pretrained on 160K hours of
   diverse music, so the marginal value of adding more piano audio is lower than
   for symbolic encoders which start from scratch.
2. **Audio data is expensive.** MAESTRO audio is ~200GB; downloading and processing
   it on Thunder Compute is costly relative to the expected benefit.
3. **Stage 1 is self-supervised on embeddings.** The cross-performer contrastive and
   augmentation invariance objectives work well with PercePiano's 1,202 segments.

**Decision (2026-02-24):** Proceeding with PercePiano-only Stage 1 for initial run.
MAESTRO MuQ embeddings (24K segments, 204 pieces) are available in GDrive. See
Section 5b below for the ablation cell if results warrant it.

In [9]:
A2_CONFIG = {
    'input_dim': 1024,
    'hidden_dim': 512,
    'num_labels': NUM_DIMS,
    'learning_rate': 3e-5,
    'weight_decay': 1e-5,
    'temperature': 0.07,
    'lambda_contrastive': 0.3,
    'lambda_regression': 0.5,
    'lambda_invariance': 0.5,
    'warmup_epochs': 5,
    'max_epochs': 200,
    'use_pretrained_muq': False,
}

STAGE1_EPOCHS = 50
STAGE2_EPOCHS = 150

print('Training A2: Staged Domain Adaptation')
print('=' * 50)

a2_trainers = []
a2_models = []
for fold_idx, fold in enumerate(folds):
    print(f'\nFold {fold_idx}/{len(folds)-1}')

    # Check if Stage 2 (final) is already done
    existing_s2 = find_checkpoint(CHECKPOINT_DIR, 'A2', fold_idx)
    if existing_s2:
        print(f'  Loading existing A2 checkpoint: {existing_s2.name}')
        model = MuQStagedModel.load_from_checkpoint(existing_s2, **A2_CONFIG, stage='supervised')
        a2_models.append(model)
        a2_trainers.append(None)
        continue

    # --- Stage 1: Self-supervised ---
    existing_s1 = find_checkpoint(CHECKPOINT_DIR, 'A2_stage1', fold_idx)
    if existing_s1:
        print(f'  Stage 1: loading existing checkpoint: {existing_s1.name}')
        model = MuQStagedModel.load_from_checkpoint(existing_s1, **A2_CONFIG, stage='self_supervised')
    else:
        print('  Stage 1: Self-supervised (contrastive + invariance)')
        model = MuQStagedModel(**A2_CONFIG, stage='self_supervised')
        train_loader_ss, val_loader_ss = build_dataloaders(
            fold, labels, piece_to_keys, embeddings, cache_dir,
            batch_size=_BATCH_SIZE, include_augmented=True,
        )
        train_model(
            model, train_loader_ss, val_loader_ss,
            'A2_stage1', fold_idx,
            checkpoint_dir=CHECKPOINT_DIR,
            max_epochs=STAGE1_EPOCHS,
            patience=10,
            accumulate_grad_batches=_ACCUM_BATCHES,
        )

    # --- Stage 2: Supervised ---
    print('  Stage 2: Supervised (ranking + regression)')
    model.switch_to_supervised()
    train_loader, val_loader = build_dataloaders(
        fold, labels, piece_to_keys, embeddings, cache_dir,
        batch_size=_BATCH_SIZE,
    )
    trainer_s2 = train_model(
        model, train_loader, val_loader,
        'A2', fold_idx,
        checkpoint_dir=CHECKPOINT_DIR,
        max_epochs=STAGE2_EPOCHS,
        patience=10,
        accumulate_grad_batches=_ACCUM_BATCHES,
    )
    a2_trainers.append(trainer_s2)
    a2_models.append(trainer_s2.lightning_module)

    best_val = trainer_s2.callback_metrics.get('val_loss', float('inf'))
    best_acc = trainer_s2.callback_metrics.get('val_pairwise_acc', 0.0)
    print(f'  Best val_loss={best_val:.4f}, val_pairwise_acc={best_acc:.4f}')

Training A2: Staged Domain Adaptation

Fold 0/3
  Loading existing A2 checkpoint: epoch=6-val_loss=0.4693.ckpt

Fold 1/3
  Loading existing A2 checkpoint: epoch=4-val_loss=0.5396.ckpt

Fold 2/3
  Loading existing A2 checkpoint: epoch=5-val_loss=0.4822.ckpt

Fold 3/3
  Loading existing A2 checkpoint: epoch=3-val_loss=0.5327.ckpt


### 5b. Ablation: A2 Stage 1 with MAESTRO Data

Optional. Re-runs A2 Stage 1 using PercePiano + MAESTRO cross-performer
contrastive data (~25x more segments). Set `RUN_MAESTRO_ABLATION = True`
to enable after reviewing baseline A2 results.

MAESTRO adds 24,321 segments across 204 pieces (vs PercePiano's 964
segments / 61 pieces). All MAESTRO pieces have >=2 performer recordings,
providing richer cross-performer contrastive signal for Stage 1.

In [ ]:
RUN_MAESTRO_ABLATION = True

if RUN_MAESTRO_ABLATION:
    import gc
    from torch.utils.data import ConcatDataset

    # -- Free prior models to reclaim memory before MAESTRO training --
    # (Evaluation cells reload from checkpoints; no data is lost.)
    a1_models, a1_trainers = [], []
    a2_models, a2_trainers = [], []
    gc.collect()
    if hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
        torch.mps.empty_cache()

    # -- Sync MAESTRO MuQ embeddings + contrastive mapping from GDrive (skip if local) --
    maestro_emb_dir = DATA_DIR / 'maestro_cache' / 'muq_embeddings'
    maestro_mapping_path = DATA_DIR / 'maestro_cache' / 'contrastive_mapping.json'

    if not any(maestro_emb_dir.glob('*.pt')):
        maestro_emb_dir.mkdir(parents=True, exist_ok=True)
        subprocess.run([
            'rclone', 'sync',
            'gdrive:crescendai_data/model_improvement/data/maestro_cache/muq_embeddings',
            str(maestro_emb_dir), '--progress',
        ], check=True)
    else:
        print(f'MAESTRO embeddings already cached ({len(list(maestro_emb_dir.glob("*.pt")))} files)')

    if not maestro_mapping_path.exists():
        subprocess.run([
            'rclone', 'copy',
            'gdrive:crescendai_data/model_improvement/data/maestro_cache/contrastive_mapping.json',
            str(maestro_mapping_path.parent), '--progress',
        ], check=True)
    else:
        print('MAESTRO contrastive mapping already cached')

    with open(maestro_mapping_path) as f:
        maestro_contrastive = json.load(f)

    # -- Build MAESTRO dataset (imported from data.py, picklable for multiprocessing) --
    piece_id_offset = len(piece_to_keys)
    maestro_ds = MaestroContrastiveDataset(
        maestro_emb_dir, maestro_contrastive, piece_id_offset=piece_id_offset,
        max_frames=500,
    )

    # On macOS (spawn), ConcatDataset with in-memory AudioSegmentDataset
    # gets pickled into each worker -- OOM with 2+ workers on 32GB.
    # Use num_workers=0 for Stage 1 combined loader.
    _s1_workers = _NUM_WORKERS if torch.cuda.is_available() else 0

    print('Training A2 (MAESTRO ablation): Staged Domain Adaptation')
    print('=' * 60)

    a2_maestro_trainers = []
    a2_maestro_models = []
    for fold_idx, fold in enumerate(folds):
        print(f'\nFold {fold_idx}/{len(folds)-1}')

        # Check if Stage 2 (final) is already done
        existing_s2 = find_checkpoint(CHECKPOINT_DIR, 'A2_maestro', fold_idx)
        if existing_s2:
            print(f'  Loading existing A2_maestro checkpoint: {existing_s2.name}')
            model = MuQStagedModel.load_from_checkpoint(existing_s2, **A2_CONFIG, stage='supervised')
            a2_maestro_models.append(model)
            a2_maestro_trainers.append(None)
            continue

        # Stage 1: check for existing checkpoint
        existing_s1 = find_checkpoint(CHECKPOINT_DIR, 'A2_maestro_stage1', fold_idx)
        if existing_s1:
            print(f'  Stage 1: loading existing checkpoint: {existing_s1.name}')
            model = MuQStagedModel.load_from_checkpoint(existing_s1, **A2_CONFIG, stage='self_supervised')
        else:
            print('  Stage 1: Self-supervised (PercePiano + MAESTRO)')
            model = MuQStagedModel(**A2_CONFIG, stage='self_supervised')

            pp_train_ds = AudioSegmentDataset(
                embeddings=embeddings,
                piece_to_keys=piece_to_keys,
                keys=fold['train'],
                max_frames=500,
            )
            combined_ds = ConcatDataset([pp_train_ds, maestro_ds])

            pp_val_ds = AudioSegmentDataset(
                embeddings=embeddings,
                piece_to_keys=piece_to_keys,
                keys=fold['val'],
                max_frames=500,
            )

            train_loader_s1 = DataLoader(
                combined_ds, batch_size=_BATCH_SIZE, shuffle=True,
                collate_fn=stage1_collate_fn, num_workers=_s1_workers,
            )
            val_loader_s1 = DataLoader(
                pp_val_ds, batch_size=_BATCH_SIZE, shuffle=False,
                collate_fn=stage1_collate_fn, num_workers=_s1_workers,
            )

            train_model(
                model, train_loader_s1, val_loader_s1,
                'A2_maestro_stage1', fold_idx,
                checkpoint_dir=CHECKPOINT_DIR,
                max_epochs=STAGE1_EPOCHS,
                accumulate_grad_batches=_ACCUM_BATCHES,
            )

        # Stage 2: Supervised on PercePiano only (unchanged from baseline A2)
        print('  Stage 2: Supervised (PercePiano only)')
        model.switch_to_supervised()

        train_loader, val_loader = build_dataloaders(
            fold, labels, piece_to_keys, embeddings, cache_dir,
            batch_size=_BATCH_SIZE,
        )

        trainer_s2 = train_model(
            model, train_loader, val_loader,
            'A2_maestro', fold_idx,
            checkpoint_dir=CHECKPOINT_DIR,
            max_epochs=STAGE2_EPOCHS,
            accumulate_grad_batches=_ACCUM_BATCHES,
        )
        a2_maestro_trainers.append(trainer_s2)
        a2_maestro_models.append(trainer_s2.lightning_module)

        best_val = trainer_s2.callback_metrics.get('val_loss', float('inf'))
        best_acc = trainer_s2.callback_metrics.get('val_pairwise_acc', 0.0)
        print(f'  Best val_loss={best_val:.4f}, val_pairwise_acc={best_acc:.4f}')

    print(f'\nA2 MAESTRO ablation complete. Stored in a2_maestro_trainers.')
    print('Compare with baseline A2 in sections 9-12.')
else:
    a2_maestro_trainers = []
    a2_maestro_models = []
    print('MAESTRO ablation skipped (RUN_MAESTRO_ABLATION = False).')
    print('Set RUN_MAESTRO_ABLATION = True above to enable after reviewing baseline A2 results.')

GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
/Users/jdhiman/Documents/crescendai/model/.venv/lib/python3.12/site-packages/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:76: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `pytorch_lightning` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default

  | Name            | Type                     | Params | Mode 
---------------------------------------------------------------------
0 | attn            | Sequential               | 262 K  | train
1 | encoder         | Sequential               | 789 K  | train
2 | projection      | Sequential               | 393 K  | tra

MAESTRO embeddings already cached (24321 files)
MAESTRO contrastive mapping already cached
MaestroContrastiveDataset: 12181 segments, 204 pieces (offset=61)
Training A2 (MAESTRO ablation): Staged Domain Adaptation

Fold 0/3
  Stage 1: Self-supervised (PercePiano + MAESTRO)


/Users/jdhiman/Documents/crescendai/model/.venv/lib/python3.12/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:433: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=9` in the `DataLoader` to improve performance.
/Users/jdhiman/Documents/crescendai/model/.venv/lib/python3.12/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:433: The 'val_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=9` in the `DataLoader` to improve performance.


Training: |          | 0/? [00:00<?, ?it/s]

  batch 50/3186 loss=0.1251
  batch 100/3186 loss=0.3978
  batch 150/3186 loss=0.1796
  batch 200/3186 loss=0.2362
  batch 250/3186 loss=0.1945
  batch 300/3186 loss=0.1988
  batch 350/3186 loss=0.3429
  batch 400/3186 loss=0.2948
  batch 450/3186 loss=0.2521
  batch 500/3186 loss=0.1842
  batch 550/3186 loss=0.3104
  batch 600/3186 loss=0.1863
  batch 650/3186 loss=0.2227
  batch 700/3186 loss=0.1602
  batch 750/3186 loss=0.3188
  batch 800/3186 loss=0.2146
  batch 850/3186 loss=0.2046
  batch 900/3186 loss=0.2418
  batch 950/3186 loss=0.3554
  batch 1000/3186 loss=0.5148
  batch 1050/3186 loss=0.2584
  batch 1100/3186 loss=0.1260
  batch 1150/3186 loss=0.2638
  batch 1200/3186 loss=0.2753
  batch 1250/3186 loss=0.3456
  batch 1300/3186 loss=0.1940
  batch 1350/3186 loss=0.2372
  batch 1400/3186 loss=0.1947
  batch 1450/3186 loss=0.1421
  batch 1500/3186 loss=0.0962
  batch 1550/3186 loss=0.2894
  batch 1600/3186 loss=0.2345
  batch 1650/3186 loss=0.1728
  batch 1700/3186 loss=0.1384


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 0: {'train_loss': '0.5366', 'train_contrast_loss': '0.3471', 'train_inv_loss': '0.3791', 'val_loss': '0.9945', 'val_contrast_loss': '0.9945', 'val_inv_loss': '0.0000'} | RAM: 3.1GB
  batch 50/3186 loss=0.4533
  batch 100/3186 loss=0.1825
  batch 150/3186 loss=0.2714
  batch 200/3186 loss=0.2403
  batch 250/3186 loss=0.1125
  batch 300/3186 loss=0.0508
  batch 350/3186 loss=0.0997
  batch 400/3186 loss=0.1395
  batch 450/3186 loss=0.0610
  batch 500/3186 loss=0.1509
  batch 550/3186 loss=0.1172
  batch 600/3186 loss=0.1067
  batch 650/3186 loss=0.0673
  batch 700/3186 loss=0.0788
  batch 750/3186 loss=0.0942
  batch 800/3186 loss=0.0975
  batch 850/3186 loss=0.0537
  batch 900/3186 loss=0.0830
  batch 950/3186 loss=0.1038
  batch 1000/3186 loss=0.1489
  batch 1050/3186 loss=0.0653
  batch 1100/3186 loss=0.0599
  batch 1150/3186 loss=0.1390
  batch 1200/3186 loss=0.1696
  batch 1250/3186 loss=0.0909
  batch 1300/3186 loss=0.0465
  batch 1350/3186 loss=0.0734
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 1: {'train_loss': '0.1539', 'train_contrast_loss': '0.0174', 'train_inv_loss': '0.2730', 'val_loss': '0.6962', 'val_contrast_loss': '0.6962', 'val_inv_loss': '0.0000'} | RAM: 3.2GB
  batch 50/3186 loss=0.0399
  batch 100/3186 loss=0.0337
  batch 150/3186 loss=0.0532
  batch 200/3186 loss=0.0319
  batch 250/3186 loss=0.0309
  batch 300/3186 loss=0.0374
  batch 350/3186 loss=0.4971
  batch 400/3186 loss=0.0318
  batch 450/3186 loss=0.0313
  batch 500/3186 loss=0.0344
  batch 550/3186 loss=0.0271
  batch 600/3186 loss=0.0279
  batch 650/3186 loss=0.0338
  batch 700/3186 loss=0.0371
  batch 750/3186 loss=0.0309
  batch 800/3186 loss=0.0216
  batch 850/3186 loss=0.0283
  batch 900/3186 loss=0.0279
  batch 950/3186 loss=0.2818
  batch 1000/3186 loss=0.0280
  batch 1050/3186 loss=0.0236
  batch 1100/3186 loss=0.0337
  batch 1150/3186 loss=0.0220
  batch 1200/3186 loss=0.0373
  batch 1250/3186 loss=0.0348
  batch 1300/3186 loss=0.0354
  batch 1350/3186 loss=0.0319
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 2: {'train_loss': '0.0809', 'train_contrast_loss': '0.0352', 'train_inv_loss': '0.0914', 'val_loss': '0.7927', 'val_contrast_loss': '0.7927', 'val_inv_loss': '0.0000'} | RAM: 3.3GB
  batch 50/3186 loss=0.0204
  batch 100/3186 loss=0.0141
  batch 150/3186 loss=0.0311
  batch 200/3186 loss=0.0125
  batch 250/3186 loss=0.0130
  batch 300/3186 loss=0.0118
  batch 350/3186 loss=0.1478
  batch 400/3186 loss=0.0165
  batch 450/3186 loss=0.0116
  batch 500/3186 loss=0.0162
  batch 550/3186 loss=0.0089
  batch 600/3186 loss=0.0093
  batch 650/3186 loss=0.0106
  batch 700/3186 loss=0.0105
  batch 750/3186 loss=0.0330
  batch 800/3186 loss=0.0084
  batch 850/3186 loss=0.0112
  batch 900/3186 loss=0.0097
  batch 950/3186 loss=0.0129
  batch 1000/3186 loss=0.0094
  batch 1050/3186 loss=0.0092
  batch 1100/3186 loss=0.0140
  batch 1150/3186 loss=0.0138
  batch 1200/3186 loss=0.0119
  batch 1250/3186 loss=0.0134
  batch 1300/3186 loss=0.0099
  batch 1350/3186 loss=0.0072
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 3: {'train_loss': '0.0239', 'train_contrast_loss': '0.0005', 'train_inv_loss': '0.0469', 'val_loss': '0.8760', 'val_contrast_loss': '0.8760', 'val_inv_loss': '0.0000'} | RAM: 3.3GB
  batch 50/3186 loss=0.0074
  batch 100/3186 loss=0.0062
  batch 150/3186 loss=0.0059
  batch 200/3186 loss=0.0128
  batch 250/3186 loss=0.0067
  batch 300/3186 loss=0.0071
  batch 350/3186 loss=0.0063
  batch 400/3186 loss=0.0138
  batch 450/3186 loss=0.0065
  batch 500/3186 loss=0.0161
  batch 550/3186 loss=0.0303
  batch 600/3186 loss=0.0067
  batch 650/3186 loss=0.0071
  batch 700/3186 loss=0.2167
  batch 750/3186 loss=0.0052
  batch 800/3186 loss=0.0061
  batch 850/3186 loss=0.0080
  batch 900/3186 loss=0.3473
  batch 950/3186 loss=0.0059
  batch 1000/3186 loss=0.0074
  batch 1050/3186 loss=0.0064
  batch 1100/3186 loss=0.0055
  batch 1150/3186 loss=0.0056
  batch 1200/3186 loss=0.0065
  batch 1250/3186 loss=0.0073
  batch 1300/3186 loss=0.0045
  batch 1350/3186 loss=0.0045
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 4: {'train_loss': '0.0209', 'train_contrast_loss': '0.0024', 'train_inv_loss': '0.0370', 'val_loss': '0.9744', 'val_contrast_loss': '0.9744', 'val_inv_loss': '0.0000'} | RAM: 3.4GB
  batch 50/3186 loss=0.0054
  batch 100/3186 loss=0.0053
  batch 150/3186 loss=0.0086
  batch 200/3186 loss=0.0043
  batch 250/3186 loss=0.0040
  batch 300/3186 loss=0.0040
  batch 350/3186 loss=0.0037
  batch 400/3186 loss=0.0049
  batch 450/3186 loss=0.0041
  batch 500/3186 loss=0.0050
  batch 550/3186 loss=0.0041
  batch 600/3186 loss=0.0082
  batch 650/3186 loss=0.0032
  batch 700/3186 loss=0.0036
  batch 750/3186 loss=0.0036
  batch 800/3186 loss=0.0037
  batch 850/3186 loss=0.0035
  batch 900/3186 loss=0.0136
  batch 950/3186 loss=0.0045
  batch 1000/3186 loss=0.0046
  batch 1050/3186 loss=0.0041
  batch 1100/3186 loss=0.0047
  batch 1150/3186 loss=0.0056
  batch 1200/3186 loss=0.0043
  batch 1250/3186 loss=0.0135
  batch 1300/3186 loss=0.0056
  batch 1350/3186 loss=0.0045
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 5: {'train_loss': '0.0141', 'train_contrast_loss': '0.0009', 'train_inv_loss': '0.0263', 'val_loss': '0.9669', 'val_contrast_loss': '0.9669', 'val_inv_loss': '0.0000'} | RAM: 3.5GB
  batch 50/3186 loss=0.0036
  batch 100/3186 loss=0.0065
  batch 150/3186 loss=0.0036
  batch 200/3186 loss=0.0026
  batch 250/3186 loss=0.0033
  batch 300/3186 loss=0.0035
  batch 350/3186 loss=0.0037
  batch 400/3186 loss=0.0035
  batch 450/3186 loss=0.0033
  batch 500/3186 loss=0.0025
  batch 550/3186 loss=0.0031
  batch 600/3186 loss=0.0029
  batch 650/3186 loss=0.0029
  batch 700/3186 loss=0.0026
  batch 750/3186 loss=0.0046
  batch 800/3186 loss=0.0050
  batch 850/3186 loss=0.0034
  batch 900/3186 loss=0.0029
  batch 950/3186 loss=0.0031
  batch 1000/3186 loss=0.0035
  batch 1050/3186 loss=0.0032
  batch 1100/3186 loss=0.0062
  batch 1150/3186 loss=0.0035
  batch 1200/3186 loss=0.0031
  batch 1250/3186 loss=0.0032
  batch 1300/3186 loss=0.0028
  batch 1350/3186 loss=0.0028
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 6: {'train_loss': '0.0117', 'train_contrast_loss': '0.0020', 'train_inv_loss': '0.0193', 'val_loss': '0.9266', 'val_contrast_loss': '0.9266', 'val_inv_loss': '0.0000'} | RAM: 3.5GB
  batch 50/3186 loss=0.0025
  batch 100/3186 loss=0.0030
  batch 150/3186 loss=0.0034
  batch 200/3186 loss=0.0044
  batch 250/3186 loss=0.0022
  batch 300/3186 loss=0.0025
  batch 350/3186 loss=0.0091
  batch 400/3186 loss=0.0040
  batch 450/3186 loss=0.0038
  batch 500/3186 loss=0.0024
  batch 550/3186 loss=0.0157
  batch 600/3186 loss=0.0032
  batch 650/3186 loss=0.0024
  batch 700/3186 loss=0.0034
  batch 750/3186 loss=0.0029
  batch 800/3186 loss=0.0027
  batch 850/3186 loss=0.0031
  batch 900/3186 loss=0.0022
  batch 950/3186 loss=0.0030
  batch 1000/3186 loss=0.0024
  batch 1050/3186 loss=0.0025
  batch 1100/3186 loss=0.0022
  batch 1150/3186 loss=0.0024
  batch 1200/3186 loss=0.0021
  batch 1250/3186 loss=0.0026
  batch 1300/3186 loss=0.0025
  batch 1350/3186 loss=0.0026
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 7: {'train_loss': '0.0089', 'train_contrast_loss': '0.0001', 'train_inv_loss': '0.0176', 'val_loss': '0.8108', 'val_contrast_loss': '0.8108', 'val_inv_loss': '0.0000'} | RAM: 3.6GB
  batch 50/3186 loss=0.0023
  batch 100/3186 loss=0.0024
  batch 150/3186 loss=0.0028
  batch 200/3186 loss=0.0027
  batch 250/3186 loss=0.0020
  batch 300/3186 loss=0.0023
  batch 350/3186 loss=0.0749
  batch 400/3186 loss=0.0020
  batch 450/3186 loss=0.0024
  batch 500/3186 loss=0.0030
  batch 550/3186 loss=0.0023
  batch 600/3186 loss=0.0097
  batch 650/3186 loss=0.0017
  batch 700/3186 loss=0.0027
  batch 750/3186 loss=0.0020
  batch 800/3186 loss=0.0031
  batch 850/3186 loss=0.0024
  batch 900/3186 loss=0.0026
  batch 950/3186 loss=0.0021
  batch 1000/3186 loss=0.0030
  batch 1050/3186 loss=0.0023
  batch 1100/3186 loss=0.0022
  batch 1150/3186 loss=0.0021
  batch 1200/3186 loss=0.0144
  batch 1250/3186 loss=0.0026
  batch 1300/3186 loss=0.0021
  batch 1350/3186 loss=0.0017
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 8: {'train_loss': '0.0087', 'train_contrast_loss': '0.0011', 'train_inv_loss': '0.0152', 'val_loss': '0.9425', 'val_contrast_loss': '0.9425', 'val_inv_loss': '0.0000'} | RAM: 3.7GB
  batch 50/3186 loss=0.0021
  batch 100/3186 loss=0.0017
  batch 150/3186 loss=0.0039
  batch 200/3186 loss=0.0019
  batch 250/3186 loss=0.0021
  batch 300/3186 loss=0.0020
  batch 350/3186 loss=0.0020
  batch 400/3186 loss=0.0019
  batch 450/3186 loss=0.0017
  batch 500/3186 loss=0.0016
  batch 550/3186 loss=0.0026
  batch 600/3186 loss=0.0036
  batch 650/3186 loss=0.0018
  batch 700/3186 loss=0.0020
  batch 750/3186 loss=0.0025
  batch 800/3186 loss=0.0023
  batch 850/3186 loss=0.0020
  batch 900/3186 loss=0.0018
  batch 950/3186 loss=0.0044
  batch 1000/3186 loss=0.0018
  batch 1050/3186 loss=0.0021
  batch 1100/3186 loss=0.0024
  batch 1150/3186 loss=0.0020
  batch 1200/3186 loss=0.0025
  batch 1250/3186 loss=0.0017
  batch 1300/3186 loss=0.0017
  batch 1350/3186 loss=0.0017
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 9: {'train_loss': '0.0125', 'train_contrast_loss': '0.0062', 'train_inv_loss': '0.0125', 'val_loss': '0.8591', 'val_contrast_loss': '0.8591', 'val_inv_loss': '0.0000'} | RAM: 3.8GB
  batch 50/3186 loss=0.0021
  batch 100/3186 loss=0.0017
  batch 150/3186 loss=0.0024
  batch 200/3186 loss=0.0017
  batch 250/3186 loss=0.0015
  batch 300/3186 loss=0.0020
  batch 350/3186 loss=0.0015
  batch 400/3186 loss=0.0045
  batch 450/3186 loss=0.0016
  batch 500/3186 loss=0.0026
  batch 550/3186 loss=0.0018
  batch 600/3186 loss=0.0018
  batch 650/3186 loss=0.0018
  batch 700/3186 loss=0.0018
  batch 750/3186 loss=0.0187
  batch 800/3186 loss=0.0034
  batch 850/3186 loss=0.0017
  batch 900/3186 loss=0.0020
  batch 950/3186 loss=0.0024
  batch 1000/3186 loss=0.0017
  batch 1050/3186 loss=0.0019
  batch 1100/3186 loss=0.0017
  batch 1150/3186 loss=0.0017
  batch 1200/3186 loss=0.0018
  batch 1250/3186 loss=0.0016
  batch 1300/3186 loss=0.0016
  batch 1350/3186 loss=0.0088
  batch 1400/3186 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 10: {'train_loss': '0.0053', 'train_contrast_loss': '0.0001', 'train_inv_loss': '0.0105', 'val_loss': '0.7730', 'val_contrast_loss': '0.7730', 'val_inv_loss': '0.0000'} | RAM: 3.8GB
  batch 50/3186 loss=0.0015
  batch 100/3186 loss=0.0016
  batch 150/3186 loss=0.0025
  batch 200/3186 loss=0.0020
  batch 250/3186 loss=0.0016
  batch 300/3186 loss=0.0021
  batch 350/3186 loss=0.0017
  batch 400/3186 loss=0.0013
  batch 450/3186 loss=0.0017
  batch 500/3186 loss=0.0025
  batch 550/3186 loss=0.0019
  batch 600/3186 loss=0.0015
  batch 650/3186 loss=0.0015
  batch 700/3186 loss=0.0015
  batch 750/3186 loss=0.0014
  batch 800/3186 loss=0.0015
  batch 850/3186 loss=0.0016
  batch 900/3186 loss=0.0017
  batch 950/3186 loss=0.0020
  batch 1000/3186 loss=0.0013
  batch 1050/3186 loss=0.0013
  batch 1100/3186 loss=0.0014
  batch 1150/3186 loss=0.0013
  batch 1200/3186 loss=0.0022
  batch 1250/3186 loss=0.0016
  batch 1300/3186 loss=0.0017
  batch 1350/3186 loss=0.0018
  batch 1400/3186 loss

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 11: {'train_loss': '0.0060', 'train_contrast_loss': '0.0006', 'train_inv_loss': '0.0107', 'val_loss': '0.7552', 'val_contrast_loss': '0.7552', 'val_inv_loss': '0.0000'} | RAM: 3.9GB


GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs

  | Name            | Type                     | Params | Mode 
---------------------------------------------------------------------
0 | attn            | Sequential               | 262 K  | train
1 | encoder         | Sequential               | 789 K  | train
2 | projection      | Sequential               | 393 K  | train
3 | comparator      | Sequential               | 1.2 M  | train
4 | ranking_heads   | ModuleList               | 1.5 K  | train
5 | regression_head | Sequential               | 132 K  | train
6 | ranking_loss    | DimensionWiseRankingLoss | 0      | train
---------------------------------------------------------------------
2.8 M     Trainable params
0         Non-trainable params
2.8 M     Total params
11.048    Total estimated model params size (MB)
39        Modules in train mode
0         Modules in eval mode


  Stage 2: Supervised (PercePiano only)


Training: |          | 0/? [00:00<?, ?it/s]

  batch 50/816 loss=0.1964
  batch 100/816 loss=0.4021
  batch 150/816 loss=0.3551
  batch 200/816 loss=0.2999
  batch 250/816 loss=0.3522
  batch 300/816 loss=0.3193
  batch 350/816 loss=0.2986
  batch 400/816 loss=0.3050
  batch 450/816 loss=0.4199
  batch 500/816 loss=0.3056
  batch 550/816 loss=0.5197
  batch 600/816 loss=0.2679
  batch 650/816 loss=0.3795
  batch 700/816 loss=0.4857
  batch 750/816 loss=0.2512
  batch 800/816 loss=0.4667


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 0: {'train_loss': '0.6222', 'train_rank_loss': '0.6195', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0054', 'val_loss': '0.6952', 'val_pairwise_acc': '0.5057'} | RAM: 5.6GB
  batch 50/816 loss=0.2632
  batch 100/816 loss=0.2561
  batch 150/816 loss=0.3644
  batch 200/816 loss=0.3247
  batch 250/816 loss=0.1825
  batch 300/816 loss=0.2896
  batch 350/816 loss=0.2619
  batch 400/816 loss=0.2400
  batch 450/816 loss=0.2971
  batch 500/816 loss=0.2830
  batch 550/816 loss=0.2537
  batch 600/816 loss=0.2569
  batch 650/816 loss=0.2404
  batch 700/816 loss=0.2834
  batch 750/816 loss=0.2649
  batch 800/816 loss=0.2127


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 1: {'train_loss': '0.6207', 'train_rank_loss': '0.6160', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0093', 'val_loss': '0.6834', 'val_pairwise_acc': '0.5186'} | RAM: 5.8GB
  batch 50/816 loss=0.2642
  batch 100/816 loss=0.2019
  batch 150/816 loss=0.2816
  batch 200/816 loss=0.3082
  batch 250/816 loss=0.2894
  batch 300/816 loss=0.2175
  batch 350/816 loss=0.2611
  batch 400/816 loss=0.2277
  batch 450/816 loss=0.1986
  batch 500/816 loss=0.2544
  batch 550/816 loss=0.2066
  batch 600/816 loss=0.2251
  batch 650/816 loss=0.2368
  batch 700/816 loss=0.2313
  batch 750/816 loss=0.2758
  batch 800/816 loss=0.2209


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 2: {'train_loss': '0.4773', 'train_rank_loss': '0.4742', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0062', 'val_loss': '0.6360', 'val_pairwise_acc': '0.6657'} | RAM: 5.9GB
  batch 50/816 loss=0.2495
  batch 100/816 loss=0.2360
  batch 150/816 loss=0.1612
  batch 200/816 loss=0.2435
  batch 250/816 loss=0.1858
  batch 300/816 loss=0.2720
  batch 350/816 loss=0.2123
  batch 400/816 loss=0.2467
  batch 450/816 loss=0.2007
  batch 500/816 loss=0.2670
  batch 550/816 loss=0.1593
  batch 600/816 loss=0.1524
  batch 650/816 loss=0.2211
  batch 700/816 loss=0.2211
  batch 750/816 loss=0.1711
  batch 800/816 loss=0.2027


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 3: {'train_loss': '0.3353', 'train_rank_loss': '0.3327', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0052', 'val_loss': '0.5502', 'val_pairwise_acc': '0.7189'} | RAM: 5.9GB
  batch 50/816 loss=0.1835
  batch 100/816 loss=0.1838
  batch 150/816 loss=0.1536
  batch 200/816 loss=0.1892
  batch 250/816 loss=0.2040
  batch 300/816 loss=0.2895
  batch 350/816 loss=0.1421
  batch 400/816 loss=0.1403
  batch 450/816 loss=0.2365
  batch 500/816 loss=0.1677
  batch 550/816 loss=0.1484
  batch 600/816 loss=0.1343
  batch 650/816 loss=0.0850
  batch 700/816 loss=0.2075
  batch 750/816 loss=0.1257
  batch 800/816 loss=0.1674


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 4: {'train_loss': '0.2294', 'train_rank_loss': '0.2220', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0148', 'val_loss': '0.5070', 'val_pairwise_acc': '0.7532'} | RAM: 6.0GB
  batch 50/816 loss=0.1088
  batch 100/816 loss=0.1183
  batch 150/816 loss=0.1596
  batch 200/816 loss=0.1829
  batch 250/816 loss=0.1069
  batch 300/816 loss=0.1125
  batch 350/816 loss=0.2283
  batch 400/816 loss=0.1542
  batch 450/816 loss=0.1030
  batch 500/816 loss=0.1740
  batch 550/816 loss=0.2697
  batch 600/816 loss=0.1862
  batch 650/816 loss=0.1265
  batch 700/816 loss=0.1619
  batch 750/816 loss=0.1303
  batch 800/816 loss=0.0940


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 5: {'train_loss': '0.4090', 'train_rank_loss': '0.4071', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0038', 'val_loss': '0.4769', 'val_pairwise_acc': '0.7681'} | RAM: 6.1GB
  batch 50/816 loss=0.1074
  batch 100/816 loss=0.1245
  batch 150/816 loss=0.0937
  batch 200/816 loss=0.1337
  batch 250/816 loss=0.1201
  batch 300/816 loss=0.1647
  batch 350/816 loss=0.1574
  batch 400/816 loss=0.0970
  batch 450/816 loss=0.0881
  batch 500/816 loss=0.2143
  batch 550/816 loss=0.0909
  batch 600/816 loss=0.0780
  batch 650/816 loss=0.1196
  batch 700/816 loss=0.1617
  batch 750/816 loss=0.1431
  batch 800/816 loss=0.1708


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 6: {'train_loss': '0.1278', 'train_rank_loss': '0.1274', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0009', 'val_loss': '0.4565', 'val_pairwise_acc': '0.7854'} | RAM: 6.1GB
  batch 50/816 loss=0.1266
  batch 100/816 loss=0.0968
  batch 150/816 loss=0.1597
  batch 200/816 loss=0.0775
  batch 250/816 loss=0.1083
  batch 300/816 loss=0.0994
  batch 350/816 loss=0.1148
  batch 400/816 loss=0.1269
  batch 450/816 loss=0.1093
  batch 500/816 loss=0.0946
  batch 550/816 loss=0.0332
  batch 600/816 loss=0.0813
  batch 650/816 loss=0.0947
  batch 700/816 loss=0.0592
  batch 750/816 loss=0.0675
  batch 800/816 loss=0.0372


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 7: {'train_loss': '0.0534', 'train_rank_loss': '0.0522', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0024', 'val_loss': '0.4545', 'val_pairwise_acc': '0.7858'} | RAM: 6.2GB
  batch 50/816 loss=0.1487
  batch 100/816 loss=0.1490
  batch 150/816 loss=0.1000
  batch 200/816 loss=0.0957
  batch 250/816 loss=0.1030
  batch 300/816 loss=0.1221
  batch 350/816 loss=0.1099
  batch 400/816 loss=0.0540
  batch 450/816 loss=0.1315
  batch 500/816 loss=0.0502
  batch 550/816 loss=0.0966
  batch 600/816 loss=0.1168
  batch 650/816 loss=0.0390
  batch 700/816 loss=0.1220
  batch 750/816 loss=0.1092
  batch 800/816 loss=0.0722


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 8: {'train_loss': '0.3755', 'train_rank_loss': '0.3735', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0039', 'val_loss': '0.4630', 'val_pairwise_acc': '0.7851'} | RAM: 6.2GB
  batch 50/816 loss=0.0502
  batch 100/816 loss=0.1191
  batch 150/816 loss=0.0404
  batch 200/816 loss=0.1659
  batch 250/816 loss=0.0716
  batch 300/816 loss=0.1216
  batch 350/816 loss=0.1312
  batch 400/816 loss=0.1311
  batch 450/816 loss=0.2174
  batch 500/816 loss=0.0905
  batch 550/816 loss=0.1164
  batch 600/816 loss=0.0581
  batch 650/816 loss=0.0446
  batch 700/816 loss=0.0647
  batch 750/816 loss=0.1267
  batch 800/816 loss=0.0295


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 9: {'train_loss': '0.3137', 'train_rank_loss': '0.3126', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0021', 'val_loss': '0.4674', 'val_pairwise_acc': '0.7868'} | RAM: 6.3GB
  batch 50/816 loss=0.1995
  batch 100/816 loss=0.0859
  batch 150/816 loss=0.0984
  batch 200/816 loss=0.0864
  batch 250/816 loss=0.0975
  batch 300/816 loss=0.0249
  batch 350/816 loss=0.0615
  batch 400/816 loss=0.0750
  batch 450/816 loss=0.0625
  batch 500/816 loss=0.0352
  batch 550/816 loss=0.0398
  batch 600/816 loss=0.0184
  batch 650/816 loss=0.0788
  batch 700/816 loss=0.0776
  batch 750/816 loss=0.0511
  batch 800/816 loss=0.0624


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 10: {'train_loss': '0.0381', 'train_rank_loss': '0.0369', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0023', 'val_loss': '0.4885', 'val_pairwise_acc': '0.7851'} | RAM: 6.3GB
  batch 50/816 loss=0.0582
  batch 100/816 loss=0.0189
  batch 150/816 loss=0.1270
  batch 200/816 loss=0.1198
  batch 250/816 loss=0.1054
  batch 300/816 loss=0.0638
  batch 350/816 loss=0.1079
  batch 400/816 loss=0.0654
  batch 450/816 loss=0.1161
  batch 500/816 loss=0.0237
  batch 550/816 loss=0.1555
  batch 600/816 loss=0.0997
  batch 650/816 loss=0.0825
  batch 700/816 loss=0.0368
  batch 750/816 loss=0.0832
  batch 800/816 loss=0.0227


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 11: {'train_loss': '0.0168', 'train_rank_loss': '0.0161', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0014', 'val_loss': '0.5030', 'val_pairwise_acc': '0.7824'} | RAM: 6.3GB
  batch 50/816 loss=0.0591
  batch 100/816 loss=0.0539
  batch 150/816 loss=0.0473
  batch 200/816 loss=0.0466
  batch 250/816 loss=0.0690
  batch 300/816 loss=0.0578
  batch 350/816 loss=0.0668
  batch 400/816 loss=0.0426
  batch 450/816 loss=0.1324
  batch 500/816 loss=0.0544
  batch 550/816 loss=0.0871
  batch 600/816 loss=0.0313
  batch 650/816 loss=0.1331
  batch 700/816 loss=0.0185
  batch 750/816 loss=0.0973
  batch 800/816 loss=0.0782


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 12: {'train_loss': '0.5394', 'train_rank_loss': '0.5378', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0032', 'val_loss': '0.5080', 'val_pairwise_acc': '0.7838'} | RAM: 6.4GB
  batch 50/816 loss=0.0502
  batch 100/816 loss=0.0658
  batch 150/816 loss=0.0656
  batch 200/816 loss=0.0200
  batch 250/816 loss=0.0321
  batch 300/816 loss=0.0473
  batch 350/816 loss=0.0235
  batch 400/816 loss=0.0298
  batch 450/816 loss=0.1428
  batch 500/816 loss=0.1000
  batch 550/816 loss=0.0335
  batch 600/816 loss=0.0501
  batch 650/816 loss=0.1953
  batch 700/816 loss=0.1732
  batch 750/816 loss=0.0322
  batch 800/816 loss=0.0575


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 13: {'train_loss': '0.0089', 'train_rank_loss': '0.0078', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0021', 'val_loss': '0.5335', 'val_pairwise_acc': '0.7791'} | RAM: 6.4GB
  batch 50/816 loss=0.0305
  batch 100/816 loss=0.1051
  batch 150/816 loss=0.0438
  batch 200/816 loss=0.0812
  batch 250/816 loss=0.0460
  batch 300/816 loss=0.0765
  batch 350/816 loss=0.0162
  batch 400/816 loss=0.0723
  batch 450/816 loss=0.0956
  batch 500/816 loss=0.0399
  batch 550/816 loss=0.0768
  batch 600/816 loss=0.0274
  batch 650/816 loss=0.0179
  batch 700/816 loss=0.0235
  batch 750/816 loss=0.0297
  batch 800/816 loss=0.0467


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 14: {'train_loss': '0.0295', 'train_rank_loss': '0.0286', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0018', 'val_loss': '0.5258', 'val_pairwise_acc': '0.7848'} | RAM: 6.4GB
  batch 50/816 loss=0.1189
  batch 100/816 loss=0.0302
  batch 150/816 loss=0.0819
  batch 200/816 loss=0.0163
  batch 250/816 loss=0.0294
  batch 300/816 loss=0.0271
  batch 350/816 loss=0.0258
  batch 400/816 loss=0.0899
  batch 450/816 loss=0.0586
  batch 500/816 loss=0.2166
  batch 550/816 loss=0.0309
  batch 600/816 loss=0.0447
  batch 650/816 loss=0.0673
  batch 700/816 loss=0.0514
  batch 750/816 loss=0.0385
  batch 800/816 loss=0.0852


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 15: {'train_loss': '0.0171', 'train_rank_loss': '0.0162', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0020', 'val_loss': '0.5520', 'val_pairwise_acc': '0.7844'} | RAM: 6.5GB
  batch 50/816 loss=0.1079
  batch 100/816 loss=0.0446
  batch 150/816 loss=0.0479
  batch 200/816 loss=0.0216
  batch 250/816 loss=0.0216
  batch 300/816 loss=0.0633
  batch 350/816 loss=0.0284
  batch 400/816 loss=0.0261
  batch 450/816 loss=0.0114
  batch 500/816 loss=0.0263
  batch 550/816 loss=0.0714
  batch 600/816 loss=0.0157
  batch 650/816 loss=0.1159
  batch 700/816 loss=0.0847
  batch 750/816 loss=0.1022
  batch 800/816 loss=0.0341


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 16: {'train_loss': '0.2170', 'train_rank_loss': '0.2159', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0022', 'val_loss': '0.5758', 'val_pairwise_acc': '0.7871'} | RAM: 6.5GB
  batch 50/816 loss=0.0331
  batch 100/816 loss=0.0249
  batch 150/816 loss=0.0171
  batch 200/816 loss=0.0058
  batch 250/816 loss=0.0870
  batch 300/816 loss=0.0108
  batch 350/816 loss=0.0298
  batch 400/816 loss=0.0294
  batch 450/816 loss=0.0953
  batch 500/816 loss=0.0125
  batch 550/816 loss=0.0464
  batch 600/816 loss=0.0458
  batch 650/816 loss=0.0235
  batch 700/816 loss=0.0332
  batch 750/816 loss=0.0300
  batch 800/816 loss=0.0281


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 17: {'train_loss': '0.2305', 'train_rank_loss': '0.2296', 'train_contrast_loss': '0.0000', 'train_reg_loss': '0.0018', 'val_loss': '0.5920', 'val_pairwise_acc': '0.7881'} | RAM: 6.5GB


GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs

  | Name            | Type                     | Params | Mode 
---------------------------------------------------------------------
0 | attn            | Sequential               | 262 K  | train
1 | encoder         | Sequential               | 789 K  | train
2 | projection      | Sequential               | 393 K  | train
3 | comparator      | Sequential               | 1.2 M  | train
4 | ranking_heads   | ModuleList               | 1.5 K  | train
5 | regression_head | Sequential               | 132 K  | train
6 | ranking_loss    | DimensionWiseRankingLoss | 0      | train
---------------------------------------------------------------------
2.8 M     Trainable params
0         Non-trainable params
2.8 M     Total params
11.048    Total estimated model params size (MB)
39        Modules in train mode
0         Modules in eval mode


  Best val_loss=0.5920, val_pairwise_acc=0.7881

Fold 1/3
  Stage 1: Self-supervised (PercePiano + MAESTRO)


Training: |          | 0/? [00:00<?, ?it/s]

  batch 50/3190 loss=0.1919
  batch 100/3190 loss=0.0904
  batch 150/3190 loss=0.1606
  batch 200/3190 loss=0.2538
  batch 250/3190 loss=0.2617
  batch 300/3190 loss=0.1846
  batch 350/3190 loss=0.1727
  batch 400/3190 loss=0.2050
  batch 450/3190 loss=0.1713
  batch 500/3190 loss=0.2402
  batch 550/3190 loss=0.1787
  batch 600/3190 loss=0.2369
  batch 650/3190 loss=0.1962
  batch 700/3190 loss=0.1876
  batch 750/3190 loss=0.4192
  batch 800/3190 loss=0.3009
  batch 850/3190 loss=0.2059
  batch 900/3190 loss=0.1746
  batch 950/3190 loss=0.3785
  batch 1000/3190 loss=0.1835
  batch 1050/3190 loss=0.3151
  batch 1100/3190 loss=0.2841
  batch 1150/3190 loss=0.3298
  batch 1200/3190 loss=0.2159
  batch 1250/3190 loss=0.1855
  batch 1300/3190 loss=0.1467
  batch 1350/3190 loss=0.2236
  batch 1400/3190 loss=0.2338
  batch 1450/3190 loss=0.1318
  batch 1500/3190 loss=0.3749
  batch 1550/3190 loss=0.1404
  batch 1600/3190 loss=0.1124
  batch 1650/3190 loss=0.1386
  batch 1700/3190 loss=0.2540


Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 0: {'train_loss': '0.1831', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.3662', 'val_loss': '0.9373', 'val_contrast_loss': '0.9373', 'val_inv_loss': '0.0000'} | RAM: 6.6GB
  batch 50/3190 loss=0.1833
  batch 100/3190 loss=0.1302
  batch 150/3190 loss=0.0836
  batch 200/3190 loss=0.0814
  batch 250/3190 loss=0.1101
  batch 300/3190 loss=0.0713
  batch 350/3190 loss=0.0836
  batch 400/3190 loss=0.0803
  batch 450/3190 loss=0.0776
  batch 500/3190 loss=0.1167
  batch 550/3190 loss=0.0617
  batch 600/3190 loss=0.0575
  batch 650/3190 loss=0.6655
  batch 700/3190 loss=0.0680
  batch 750/3190 loss=0.0822
  batch 800/3190 loss=0.0611
  batch 850/3190 loss=0.0601
  batch 900/3190 loss=0.0985
  batch 950/3190 loss=0.0601
  batch 1000/3190 loss=0.0531
  batch 1050/3190 loss=0.0771
  batch 1100/3190 loss=0.0598
  batch 1150/3190 loss=0.0933
  batch 1200/3190 loss=0.0618
  batch 1250/3190 loss=0.0443
  batch 1300/3190 loss=0.0463
  batch 1350/3190 loss=0.0601
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 1: {'train_loss': '0.1204', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.2409', 'val_loss': '0.6510', 'val_contrast_loss': '0.6510', 'val_inv_loss': '0.0000'} | RAM: 6.7GB
  batch 50/3190 loss=0.0376
  batch 100/3190 loss=0.0312
  batch 150/3190 loss=0.0273
  batch 200/3190 loss=0.0327
  batch 250/3190 loss=0.0330
  batch 300/3190 loss=0.0317
  batch 350/3190 loss=0.0259
  batch 400/3190 loss=0.0339
  batch 450/3190 loss=0.0278
  batch 500/3190 loss=0.0381
  batch 550/3190 loss=0.0275
  batch 600/3190 loss=0.0349
  batch 650/3190 loss=0.0239
  batch 700/3190 loss=0.0254
  batch 750/3190 loss=0.0211
  batch 800/3190 loss=0.0548
  batch 850/3190 loss=0.0243
  batch 900/3190 loss=0.0439
  batch 950/3190 loss=0.0899
  batch 1000/3190 loss=0.0388
  batch 1050/3190 loss=0.0269
  batch 1100/3190 loss=0.0404
  batch 1150/3190 loss=0.0233
  batch 1200/3190 loss=0.0276
  batch 1250/3190 loss=0.0251
  batch 1300/3190 loss=0.0256
  batch 1350/3190 loss=0.0622
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 2: {'train_loss': '0.0364', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0727', 'val_loss': '0.7776', 'val_contrast_loss': '0.7776', 'val_inv_loss': '0.0000'} | RAM: 6.8GB
  batch 50/3190 loss=0.0379
  batch 100/3190 loss=0.0100
  batch 150/3190 loss=0.0127
  batch 200/3190 loss=0.0176
  batch 250/3190 loss=0.0103
  batch 300/3190 loss=0.0110
  batch 350/3190 loss=0.0167
  batch 400/3190 loss=0.0118
  batch 450/3190 loss=0.0109
  batch 500/3190 loss=0.0117
  batch 550/3190 loss=0.0129
  batch 600/3190 loss=0.0201
  batch 650/3190 loss=0.0108
  batch 700/3190 loss=0.0090
  batch 750/3190 loss=0.5544
  batch 800/3190 loss=0.0098
  batch 850/3190 loss=0.0135
  batch 900/3190 loss=0.0095
  batch 950/3190 loss=0.0103
  batch 1000/3190 loss=0.0146
  batch 1050/3190 loss=0.0249
  batch 1100/3190 loss=0.0089
  batch 1150/3190 loss=0.0079
  batch 1200/3190 loss=0.0097
  batch 1250/3190 loss=0.0189
  batch 1300/3190 loss=0.0136
  batch 1350/3190 loss=0.0108
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 3: {'train_loss': '0.0199', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0398', 'val_loss': '0.8168', 'val_contrast_loss': '0.8168', 'val_inv_loss': '0.0000'} | RAM: 6.8GB
  batch 50/3190 loss=0.0095
  batch 100/3190 loss=0.0066
  batch 150/3190 loss=0.0069
  batch 200/3190 loss=0.0068
  batch 250/3190 loss=0.0109
  batch 300/3190 loss=0.0062
  batch 350/3190 loss=0.0056
  batch 400/3190 loss=0.0050
  batch 450/3190 loss=0.0063
  batch 500/3190 loss=0.0064
  batch 550/3190 loss=0.0061
  batch 600/3190 loss=0.0096
  batch 650/3190 loss=0.8392
  batch 700/3190 loss=0.0111
  batch 750/3190 loss=0.0613
  batch 800/3190 loss=0.0057
  batch 850/3190 loss=0.0053
  batch 900/3190 loss=0.0067
  batch 950/3190 loss=0.0075
  batch 1000/3190 loss=0.0075
  batch 1050/3190 loss=0.3028
  batch 1100/3190 loss=0.0045
  batch 1150/3190 loss=0.0054
  batch 1200/3190 loss=0.0114
  batch 1250/3190 loss=0.0051
  batch 1300/3190 loss=0.0057
  batch 1350/3190 loss=0.0063
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 4: {'train_loss': '0.0156', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0312', 'val_loss': '0.8205', 'val_contrast_loss': '0.8205', 'val_inv_loss': '0.0000'} | RAM: 6.9GB
  batch 50/3190 loss=0.0076
  batch 100/3190 loss=0.0038
  batch 150/3190 loss=0.0042
  batch 200/3190 loss=0.0043
  batch 250/3190 loss=0.0043
  batch 300/3190 loss=0.0041
  batch 350/3190 loss=0.0041
  batch 400/3190 loss=0.0179
  batch 450/3190 loss=0.0094
  batch 500/3190 loss=0.0040
  batch 550/3190 loss=0.0044
  batch 600/3190 loss=0.0037
  batch 650/3190 loss=0.0042
  batch 700/3190 loss=0.0036
  batch 750/3190 loss=0.0036
  batch 800/3190 loss=0.0042
  batch 850/3190 loss=0.0048
  batch 900/3190 loss=0.0041
  batch 950/3190 loss=0.0045
  batch 1000/3190 loss=0.0039
  batch 1050/3190 loss=0.0042
  batch 1100/3190 loss=0.0056
  batch 1150/3190 loss=0.2355
  batch 1200/3190 loss=0.0036
  batch 1250/3190 loss=0.0035
  batch 1300/3190 loss=0.0048
  batch 1350/3190 loss=0.0038
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 5: {'train_loss': '0.0111', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0222', 'val_loss': '0.8355', 'val_contrast_loss': '0.8355', 'val_inv_loss': '0.0000'} | RAM: 7.0GB
  batch 50/3190 loss=0.1124
  batch 100/3190 loss=0.0027
  batch 150/3190 loss=0.0041
  batch 200/3190 loss=0.0034
  batch 250/3190 loss=0.0036
  batch 300/3190 loss=0.0033
  batch 350/3190 loss=0.0031
  batch 400/3190 loss=0.0032
  batch 450/3190 loss=0.0045
  batch 500/3190 loss=0.0026
  batch 550/3190 loss=0.0029
  batch 600/3190 loss=0.0058
  batch 650/3190 loss=0.0029
  batch 700/3190 loss=0.0029
  batch 750/3190 loss=0.0067
  batch 800/3190 loss=0.0027
  batch 850/3190 loss=0.0029
  batch 900/3190 loss=0.0025
  batch 950/3190 loss=0.0034
  batch 1000/3190 loss=0.0157
  batch 1050/3190 loss=0.0075
  batch 1100/3190 loss=0.0041
  batch 1150/3190 loss=0.4536
  batch 1200/3190 loss=0.0037
  batch 1250/3190 loss=0.0027
  batch 1300/3190 loss=0.0028
  batch 1350/3190 loss=0.0034
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 6: {'train_loss': '0.0137', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0275', 'val_loss': '0.8019', 'val_contrast_loss': '0.8019', 'val_inv_loss': '0.0000'} | RAM: 7.1GB
  batch 50/3190 loss=0.0028
  batch 100/3190 loss=0.0027
  batch 150/3190 loss=0.0025
  batch 200/3190 loss=0.0024
  batch 250/3190 loss=0.0025
  batch 300/3190 loss=0.0072
  batch 350/3190 loss=0.2387
  batch 400/3190 loss=0.0025
  batch 450/3190 loss=0.0036
  batch 500/3190 loss=0.0033
  batch 550/3190 loss=0.0027
  batch 600/3190 loss=0.0025
  batch 650/3190 loss=0.0039
  batch 700/3190 loss=0.0026
  batch 750/3190 loss=0.0036
  batch 800/3190 loss=0.0034
  batch 850/3190 loss=0.0026
  batch 900/3190 loss=0.0025
  batch 950/3190 loss=0.0040
  batch 1000/3190 loss=0.0024
  batch 1050/3190 loss=0.0023
  batch 1100/3190 loss=0.0034
  batch 1150/3190 loss=0.0024
  batch 1200/3190 loss=0.0026
  batch 1250/3190 loss=0.0023
  batch 1300/3190 loss=0.0026
  batch 1350/3190 loss=0.0037
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 7: {'train_loss': '0.0072', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0144', 'val_loss': '0.8043', 'val_contrast_loss': '0.8043', 'val_inv_loss': '0.0000'} | RAM: 7.1GB
  batch 50/3190 loss=0.0031
  batch 100/3190 loss=0.0019
  batch 150/3190 loss=0.0026
  batch 200/3190 loss=0.0024
  batch 250/3190 loss=0.0026
  batch 300/3190 loss=0.0044
  batch 350/3190 loss=0.0023
  batch 400/3190 loss=0.3967
  batch 450/3190 loss=0.0029
  batch 500/3190 loss=0.0021
  batch 550/3190 loss=0.0022
  batch 600/3190 loss=0.0023
  batch 650/3190 loss=0.0023
  batch 700/3190 loss=0.0061
  batch 750/3190 loss=0.0035
  batch 800/3190 loss=0.0019
  batch 850/3190 loss=0.0023
  batch 900/3190 loss=0.0023
  batch 950/3190 loss=0.0021
  batch 1000/3190 loss=0.0021
  batch 1050/3190 loss=0.2781
  batch 1100/3190 loss=0.0034
  batch 1150/3190 loss=0.0019
  batch 1200/3190 loss=0.0020
  batch 1250/3190 loss=0.0077
  batch 1300/3190 loss=0.0021
  batch 1350/3190 loss=0.0030
  batch 1400/3190 loss=

Validation: |          | 0/? [00:00<?, ?it/s]

Epoch 8: {'train_loss': '0.0075', 'train_contrast_loss': '0.0000', 'train_inv_loss': '0.0150', 'val_loss': '0.7707', 'val_contrast_loss': '0.7707', 'val_inv_loss': '0.0000'} | RAM: 7.2GB
  batch 50/3190 loss=0.0026
  batch 100/3190 loss=0.0030
  batch 150/3190 loss=0.0021
  batch 200/3190 loss=0.0017
  batch 250/3190 loss=0.0119
  batch 300/3190 loss=0.0023
  batch 350/3190 loss=0.0025
  batch 400/3190 loss=0.0020
  batch 450/3190 loss=0.0026
  batch 500/3190 loss=0.0020
  batch 550/3190 loss=0.0021
  batch 600/3190 loss=0.0020
  batch 650/3190 loss=0.0020
  batch 700/3190 loss=0.0020
  batch 750/3190 loss=0.0018
  batch 800/3190 loss=0.0016
  batch 850/3190 loss=0.0016
  batch 900/3190 loss=0.3584
  batch 950/3190 loss=0.3147
  batch 1000/3190 loss=0.0034
  batch 1050/3190 loss=0.0019
  batch 1100/3190 loss=0.0020
  batch 1150/3190 loss=0.0033
  batch 1200/3190 loss=0.0020
  batch 1250/3190 loss=0.0017
  batch 1300/3190 loss=0.0020
  batch 1350/3190 loss=0.0026
  batch 1400/3190 loss=

## 6. Train A3: Full Unfreeze

In [ ]:
# -- Free prior models to reclaim memory before A3 training --
import gc
a2_maestro_models, a2_maestro_trainers = [], []
gc.collect()
if hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
    torch.mps.empty_cache()

A3_CONFIG = {
    'input_dim': 1024,
    'hidden_dim': 512,
    'num_labels': NUM_DIMS,
    'learning_rate': 3e-5,
    'weight_decay': 1e-5,
    'temperature': 0.07,
    'lambda_contrastive': 0.3,
    'lambda_regression': 0.5,
    'warmup_epochs': 5,
    'max_epochs': 200,
    'use_pretrained_muq': False,
    'unfreeze_schedule': {0: [12], 10: [11], 20: [10], 30: [9]},
    'lr_decay_factor': 0.8,
    'use_gradient_checkpointing': LOW_MEMORY,
}

print('Training A3: Full Unfreeze with Gradual Layer Unfreezing')
print('=' * 50)

a3_trainers = []
a3_models = []
for fold_idx, fold in enumerate(folds):
    print(f'\nFold {fold_idx}/{len(folds)-1}')

    existing = find_checkpoint(CHECKPOINT_DIR, 'A3', fold_idx)
    if existing:
        print(f'  Loading existing checkpoint: {existing.name}')
        model = MuQFullUnfreezeModel.load_from_checkpoint(existing, **A3_CONFIG)
        a3_models.append(model)
        a3_trainers.append(None)
        continue

    model = MuQFullUnfreezeModel(**A3_CONFIG)
    train_loader, val_loader = build_dataloaders(
        fold, labels, piece_to_keys, embeddings, cache_dir,
        batch_size=_BATCH_SIZE,
    )

    trainer = train_model(
        model, train_loader, val_loader, 'A3', fold_idx,
        checkpoint_dir=CHECKPOINT_DIR,
        max_epochs=A3_CONFIG['max_epochs'],
        patience=10,
        accumulate_grad_batches=_ACCUM_BATCHES,
    )
    a3_trainers.append(trainer)
    a3_models.append(trainer.lightning_module)

    best_val = trainer.callback_metrics.get('val_loss', float('inf'))
    best_acc = trainer.callback_metrics.get('val_pairwise_acc', 0.0)
    print(f'  Best val_loss={best_val:.4f}, val_pairwise_acc={best_acc:.4f}')

## 7. Training Summary

In [ ]:
import re

def _get_val_metrics(trainer, model_name, fold_idx):
    """Extract val metrics from trainer or checkpoint filename."""
    if trainer is not None:
        return (
            float(trainer.callback_metrics.get('val_loss', float('nan'))),
            float(trainer.callback_metrics.get('val_pairwise_acc', float('nan'))),
        )
    ckpt = find_checkpoint(CHECKPOINT_DIR, model_name, fold_idx)
    if ckpt is not None:
        match = re.search(r'val_loss=([\d.]+)', ckpt.name)
        if match:
            return float(match.group(1)), float('nan')
    return float('nan'), float('nan')

print('Training Summary')
print('=' * 60)
print(f'{"Model":<16} {"Fold":<6} {"Val Loss":<12} {"Pairwise Acc":<14}')
print('-' * 60)

trainer_groups = [('A1', a1_trainers), ('A2', a2_trainers), ('A3', a3_trainers)]
if a2_maestro_trainers:
    trainer_groups.append(('A2_maestro', a2_maestro_trainers))

for name, trainers in trainer_groups:
    for fold_idx, trainer in enumerate(trainers):
        val_loss, val_acc = _get_val_metrics(trainer, name, fold_idx)
        print(f'{name:<16} {fold_idx:<6} {val_loss:<12.4f} {val_acc:<14.4f}')

print('\nCheckpoints saved to:', CHECKPOINT_DIR)
print('Checkpoints synced to Google Drive via rclone')

## 8. Upload Final Results

In [ ]:
if IS_CLOUD:
    for model_name in ['A1', 'A2', 'A2_stage1', 'A3']:
        local = CHECKPOINT_DIR / model_name
        if local.exists():
            upload_checkpoint(local, model_name)
            print(f'Uploaded {model_name} checkpoints')
else:
    print('Local run -- skipping GDrive upload.')
    print(f'Checkpoints saved to: {CHECKPOINT_DIR.resolve()}')

## 9. Evaluate All Audio Encoders

In [ ]:
# Adapter functions for audio encoders (all three share the same API)
def audio_get_input_fn(key):
    return embeddings[key].unsqueeze(0), None

def audio_encode_fn(model, inp, mask):
    return model.encode(inp, mask)

def audio_compare_fn(model, z_a, z_b):
    return model.compare(z_a, z_b)

def audio_predict_fn(model, inp, mask):
    if hasattr(model, 'predict_scores'):
        return model.predict_scores(inp, mask)
    # A2 (MuQStagedModel) uses regression_head directly
    return model.regression_head(model.encode(inp, mask))

# Evaluate all audio encoders across all folds
all_results = {}
encoder_models = {
    'A1': a1_models,
    'A2': a2_models,
    'A3': a3_models,
}

for name, fold_models in encoder_models.items():
    print(f'\nEvaluating {name}...')
    fold_metrics = []
    for fold_idx, (fold, model) in enumerate(zip(folds, fold_models)):
        fold_res = evaluate_model(
            model, fold['val'], labels,
            get_input_fn=audio_get_input_fn,
            encode_fn=audio_encode_fn,
            compare_fn=audio_compare_fn,
            predict_fn=audio_predict_fn,
        )
        fold_metrics.append(fold_res)
        pw = fold_res.get('pairwise', 'N/A')
        r2 = fold_res.get('r2', 'N/A')
        pw_s = f'{pw:.4f}' if isinstance(pw, float) else str(pw)
        r2_s = f'{r2:.4f}' if isinstance(r2, float) else str(r2)
        print(f'  Fold {fold_idx}: pairwise={pw_s}, r2={r2_s}')

    agg = aggregate_folds(fold_metrics)
    if 'pairwise_detail' in fold_metrics[-1]:
        agg['pairwise_detail'] = fold_metrics[-1]['pairwise_detail']
    all_results[name] = agg
    print(f'  Average: pairwise={agg.get("pairwise_mean", 0):.4f}, r2={agg.get("r2_mean", 0):.4f}')

## 10. Robustness Check

In [ ]:
robustness_results = {}
val_keys = folds[-1]['val']

for name, fold_models in encoder_models.items():
    model = fold_models[-1]
    print(f'Robustness check for {name}...')
    rob = run_robustness_check(
        model, val_keys,
        get_input_fn=audio_get_input_fn,
        predict_fn=audio_predict_fn,
        noise_std=0.05,
    )
    robustness_results[name] = rob
    print(f'  pearson_r={rob["pearson_r"]:.4f}, score_drop_pct={rob["score_drop_pct"]:.1f}%')

## 11. Comparison Table + Per-Dimension Breakdown

In [ ]:
comparison = {}
for name in encoder_models:
    comparison[name] = {}
    if name in all_results:
        comparison[name]['pairwise'] = all_results[name].get('pairwise_mean', 0.0)
        comparison[name]['r2'] = all_results[name].get('r2_mean', 0.0)
    if name in robustness_results:
        comparison[name]['robustness'] = robustness_results[name]['pearson_r']
        comparison[name]['score_drop_pct'] = robustness_results[name]['score_drop_pct']

print(format_comparison_table(comparison))

# Per-dimension bar chart
fig, ax = plt.subplots(1, 1, figsize=(14, 6))
n_dims = NUM_DIMS
x = np.arange(n_dims)
width = 0.25

for i, (name, res) in enumerate(all_results.items()):
    if 'pairwise_detail' not in res:
        continue
    per_dim = res['pairwise_detail']['per_dimension']
    values = [per_dim.get(d, 0.5) for d in range(n_dims)]
    ax.bar(x + i * width, values, width, label=name, alpha=0.8)

ax.set_xticks(x + width)
ax.set_xticklabels(DIMENSIONS, rotation=45, ha='right', fontsize=8)
ax.set_ylabel('Pairwise Accuracy')
ax.set_title('Per-Dimension Pairwise Ranking Accuracy (Audio Encoders)')
ax.legend()
ax.axhline(y=0.5, color='gray', linestyle='--', alpha=0.5)
ax.set_ylim(0.3, 1.0)
plt.tight_layout()
plt.show()

## 12. Winner Selection

Selection protocol:
1. **Primary**: Highest pairwise ranking accuracy
2. **Tiebreak**: Highest R-squared on regression
3. **Veto**: Robustness score_drop_pct > 15% disqualifies the model

In [ ]:
selection_input = {}
for name, metrics in comparison.items():
    selection_input[name] = {
        'pairwise_mean': metrics.get('pairwise', 0.0),
        'r2_mean': metrics.get('r2', 0.0),
        'score_drop_pct': metrics.get('score_drop_pct', 0.0),
    }

winner = select_winner(selection_input)
print(f'\nWINNER: {winner}')
for name, metrics in comparison.items():
    marker = ' <-- WINNER' if name == winner else ''
    print(f'  {name}: pairwise={metrics.get("pairwise", 0):.4f}, r2={metrics.get("r2", 0):.4f}{marker}')